prepreduce/prepreduce.py

```diff
@@ -0,0 +1,411 @@
+    )
+
+    return local_path_for_dot_i_file
+
+
+def get_compiler_invocation(target: Path, make_stdout: str) -> Command:
+    """
+    Parse out the CC invocation from make's raw output with V=1 on
+    """
+    pattern = f".*-c -o {target}.*"
+    matches = re.findall(pattern, make_stdout)
+    ic(matches)
+    assert (
+        len(matches) == 1
+    ), "Too many (or too few) compiler invocation matches! (not your fault, it's prepreduce's fault)"
+
+    main_invocation = matches[0].split()
+    return main_invocation
+
+
+def get_clean_compiler_invocation(cc_invocation: Command) -> Command:
+    """
+    Remove all preprocessor flags as well as any relative pathing in invocation
+    """
+    to_remove = ("-I", "-D", "-Wp", "-include", "-Werror", "./", "-U")
+    rules = [(lambda rule: lambda flag: rule not in flag)(rule) for rule in to_remove]
+
+    cleaned = cc_invocation.copy()
+    for rule in rules:
+        cleaned = filter(rule, cleaned)
+
+    cleaned = list(cleaned)  # eagerly convert iterator to list
+
+    ic(cleaned)
+
+    return cleaned
+
+
+def main() -> None:
+    cli_args = parse_cli_args()
+
+    uses_clang: Optional[bool] = kernel_configured_for_clang(cli_args.path_to_linux)
+    assert uses_clang is not None
+    info(f"Using CC={'clang' if uses_clang else 'gcc'}")
+
+    ic(cli_args)
+    make_stdout = initial_make(
+        build_command=cli_args.build_command,
+        target=cli_args.target,
+        path_to_linux=cli_args.path_to_linux,
+        force_rm=cli_args.force_rm_existing_target,
+    )
+
+    ic(make_stdout)
+
+    cc_invocation = get_compiler_invocation(cli_args.target, make_stdout)
+    path_to_dot_i_file = make_dot_i_file(
+        cc_invocation, path_to_linux=cli_args.path_to_linux, output_dir=cli_args.output
+    )
+
+    clean_cc_invocation = get_clean_compiler_invocation(cc_invocation)
+    clean_cc_invocation[-1] = path_to_dot_i_file  # TODO: search for .c file
+    ic(clean_cc_invocation)
+
+    write_test_script(
+        cc_invocation=clean_cc_invocation,
+        target=path_to_dot_i_file.with_suffix(".o"),
+        force_rm=cli_args.force_rm_existing_script,
+        uses_clang=uses_clang,
+    )
+
+
+def write_test_script(
+    *, cc_invocation: Command, target: Path, force_rm: bool, uses_clang: bool
+) -> None:
+    """
+    Create a test.sh script at --output directory
+    """
+    script_path = Path.cwd() / Path("test.sh")
+    info(f"Generating {script_path}")
+    if script_path.exists():
+        double_check_removal_with_user(file=script_path, force_rm=force_rm)
+
+    output_target_idx = cc_invocation.index("-o") + 1
+    cc_invocation[output_target_idx] = target.stem + target.suffix
+
+    cc_flags = " ".join(str(x) for x in cc_invocation[1:])
+
+    script_text = textwrap.dedent(
+        f"""    #!/usr/bin/env bash
+    CC_CMD() {{
+        {'clang' if uses_clang else 'gcc'} {cc_flags}
+    }}
+    echo $(CC_CMD 2>&1) | grep \"<your test here>\"
```

Comment:
No need for `echo` here. Also, since you are using `"""`, do you need to escape the quotations?

```bash
CC_CMD 2>&1 | grep "<your test here>"
```

Reply:
The echo is nice because you can remove the `grep` (manually) and just see what the compiler is outputting. I suppose, though, that if someone is already removing the grep, they can simply add the echo themselves.

Great suggestion, fixing now.